docs/proposals/clusterconfig.md

```diff
@@ -0,0 +1,110 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Proposes a mechanism for specifying cluster-wide runtime config
+for all kube system components via a single config file. This will
+provide an easy way to specify init settings that are shared between
+system components.
+
+## Motivation
+
+Primary motivation is enabling/disabling features that are not tied to
+an API group. API groups can be selectively enabled/disabled in the
+apiserver via the `--runtime-config` commandline flag, but there is
+currently no mechanism to toggle alpha features that are controlled by
+e.g. annotations. This means the burden of controlling whether such
+features are enabled in a particular cluster is on feature implementors;
+they must either define some ad hoc mechanism for toggling (e.g. flag
+on component binary) or else toggle the feature on/off at compile time.
+
+Providing a mechanism for specifying cluster-wide config makes
+deploying non-API features simpler for implementors and cluster admins.
+
+## Design
+
+Cluster-wide runtime config will be specified in a single yaml file
+loaded from a known, default path. TBD whether default config path
+should be overwriteable via flag/env.
+
+Strawman:
+```shell
+$ cat /etc/srv/kubernetes/cluster_config.yaml
+runtimeConfig:
+  enableAppArmor: true
+  enableFooBarBaz: false
```

Comment:
> How would this work for third party extension apis (like federation, or an API group not hosted on the master)?

Not affected by this proposal I think. IIUC third party extension apis are already not managed by kube-system beyond basic resource crud and api paths. Similar answer for API groups not hosted on master.

> To argue the opposite direction here - shouldn't component config support a set of flags that indicate alpha feature enablement but are instead typed? 

Typed config does sound better than map[string]string, I'll update.

As for one config per component, that's along the lines of the `--runtime-config` alternate design below. I expect it looks like either a config file for each component, loaded from a default path, overwriteable by flag. The reason I didn't like it is because you now have to keep the toggle values consistent across components if you have a feature that touches multiple components.

> How do rolling upgrades of enabled alpha features to beta features work?

They don't. By [decision](https://github.com/kubernetes/kubernetes/blob/b829d4d4ef68e64b9b7ae42b46877ee75bb2bfd9/docs/devel/api_changes.md#alpha-beta-and-stable-versions), alpha features are not required to have an upgrade path to beta, so they should only be enabled/disabled at cluster creation time.

> How many components on average would we add these flags to, that justifies going a bit further and having a shared enablement config?

Core components are apiserver, scheduler, controller-manager, kubelet, kube-proxy. Cluster addons too, dns, at the very least. My "single feature touching multiple components" scenario could be purely imagined or uncommon, but if not, maintaining consistent enablement config across 6+ separate components sounds error prone.

> any harder than also having to reason about two config maps per process type?

I don't follow the "two config maps per process type" part of the question. Which config maps are you referring to here?